waku/v2/protocol/filter/waku_filter.go

```diff
@@ -0,0 +1,278 @@
+	// This will require us to check for various factors such as:
+	//  - which topics they track
+	//  - latency?
+	//  - default store peer?
+
+	// Selects the best peer for a given protocol
+	var peers peer.IDSlice
+	for _, peer := range wf.h.Peerstore().Peers() {
+		protocols, err := wf.h.Peerstore().SupportsProtocols(peer, string(WakuFilterProtocolId))
+		if err != nil {
+			log.Error("error obtaining the protocols supported by peers", err)
+			return nil
+		}
+
+		if len(protocols) > 0 {
+			peers = append(peers, peer)
+		}
+	}
+
+	if len(peers) >= 1 {
+		// TODO: proper heuristic here that compares peer scores and selects "best" one. For now the first peer for the given protocol is returned
+		return &peers[0]
+	}
+
+	return nil
+}
+
+func (wf *WakuFilter) onRequest(s network.Stream) {
+	defer s.Close()
+
+	filterRPCRequest := &pb.FilterRPC{}
+
+	reader := protoio.NewDelimitedReader(s, 64*1024)
+
+	err := reader.ReadMsg(filterRPCRequest)
+	if err != nil {
+		log.Error("error reading request", err)
+		return
+	}
+
+	log.Info(fmt.Sprintf("%s: Received query from %s", s.Conn().LocalPeer(), s.Conn().RemotePeer()))
+
+	if filterRPCRequest.Request != nil {
+		// We're on a full node.
+		// This is a filter request coming from a light node.
+		if filterRPCRequest.Request.Subscribe {
+			subscriber := Subscriber{peer: string(s.Conn().RemotePeer()), requestId: filterRPCRequest.RequestId, filter: *filterRPCRequest.Request}
+			wf.subscribers = append(wf.subscribers, subscriber)
+			log.Info("Full node, add a filter subscriber ", subscriber)
+		} else {
+			// TODO wf.subscribers.unsubscribeFilters(filterRPCRequest.Request, conn.peerInfo.peerId)
+		}
+	} else if filterRPCRequest.Push != nil {
+		// We're on a light node.
+		// This is a message push coming from a full node.
+
+		log.Info("Light node, received a message push ", *filterRPCRequest.Push)
+		wf.pushHandler(filterRPCRequest.RequestId, *filterRPCRequest.Push)
+	}
+
+}
+
+func NewWakuFilter(ctx context.Context, host host.Host, handler MessagePushHandler) *WakuFilter {
+	wf := new(WakuFilter)
+	wf.ctx = ctx
+	wf.MsgC = make(chan *protocol.Envelope)
+	wf.h = host
+	wf.pushHandler = handler
+
+	wf.h.SetStreamHandler(WakuFilterProtocolId, wf.onRequest)
+	go wf.FilterListener()
+
+	return wf
+}
+
+func (wf *WakuFilter) FilterListener() {
+
+	// This function is invoked for each message received
+	// on the full node in context of Waku2-Filter
+	handle := func(topic string, msg *pb.WakuMessage) error { // async
+		// trace "handle WakuFilter subscription", topic=topic, msg=msg
+
+		// Each subscriber is a light node that earlier on invoked
+		// a FilterRequest on this node
+		for _, subscriber := range wf.subscribers {
+			// TODO
+			// if subscriber.filter.pubSubTopic != "" && subscriber.filter.pubSubTopic != topic {
+			// 	log.Info("Subscriber's filter pubsubTopic does not match message topic", subscriber.filter.pubSubTopic, topic)
+			// 	continue
+			// }
```

Comment:
Can we remove this comment?
```suggestion
```

Reply:
Modified the code so that that code could be uncommented. I used hardcoded topic value in calls to that handle fn, but changed it's signature so that it accepts Envelopes instead.